src/game/client/swarm/gameui/swarm/basemodpanel.cpp

```diff
@@ -1560,39 +1567,86 @@ void CBaseModPanel::OnEvent( KeyValues *pEvent )
 			if ( !pSession )
 				return;
 
+			//KeyValues *pSysData = pSession->GetSessionSystemData();
+			//if (pSysData) KeyValuesDumpAsDevMsg(pSysData, 1, 0);
+
 			KeyValues *pSettings = pSession->GetSessionSettings();
+			if (!pSettings)
+				return;
+
+			//KeyValuesDumpAsDevMsg(pSettings, 1, 0);
+			bool dedicatedServer = false;
+			if (!Q_strcmp(pSettings->GetString("server/server", "lobby"), "dedicated"))
+				dedicatedServer = true;
 
 			KeyValues *pInfoMission = NULL;
-			// TODO:
 			KeyValues *pInfoChapter = NULL;//GetMapInfoRespectingAnyChapter( pSettings, &pInfoMission );
 
 			bool bValidMission = true;
 
-			// TODO: Check if we have the map installed by querying local mission chooser source
+			IASW_Mission_Chooser_Source *pSource = missionchooser ? missionchooser->LocalMissionSource() : NULL;
+			if (pSource)
+			{
+				pSource->Think();
+				const char *szCampaignName = pSettings->GetString("game/campaign");
+				const char *szMapName = pSettings->GetString("game/mission");
+				bValidMission = pSource->MissionExists(szMapName, false);
+				if (bValidMission)
+				{
+					Msg("Map %s is installed.\n", szMapName);
+					pInfoMission = pSource->GetMissionDetails(szMapName);
+					pInfoChapter = pSource->GetCampaignDetails(szCampaignName);
+				}
+			}
 
-			if ( bValidMission )
-				return;
+			bool bBlacklisted = false;
+			if (dedicatedServer)
+			{
+				const char* pServerAddress = pSettings->GetString("server/adronline", "0.0.0.0:0");
+				if (LoadBlackListFile(m_pServerBlackList))
```

Comment:
need to get rid of the previous blacklist keyvalues before we create the new one so we don't leak memory
(or only load once, but then we lose the hot-updates)